sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala

```diff
@@ -139,6 +143,31 @@ case class CollectSet(
 
   def this(child: Expression) = this(child, 0, 0)
 
+  /* SPARK-31500
+   * Array[Byte](BinaryType) Scala equality don't works as expected
+   * so HashSet return duplicates, we have to change types to drop
+   * this duplicates and make collect_set work as expected for this
+   * data type
+   */
+  override lazy val typeChild = child.dataType match {
+    case BinaryType => ArrayType(BinaryType)
+    case other => other
+  }
+
+  override def getValueOnUpdate(value: Any): Any = InternalRow.copyValue(value) match {
+    case v: Array[Byte] => UnsafeArrayData.fromPrimitiveArray(v)
+    case other => other
+  }
+
+  override def eval(buffer: mutable.HashSet[Any]): Any = {
+    val bufferUpdated: mutable.HashSet[Any] =
+      child.dataType match {
+        case BinaryType => buffer.map(_.asInstanceOf[UnsafeArrayData].toByteArray)
```

Comment:
The `map` call produces another set, which is not for free. We could do the following:
```scala
override def eval(buffer: mutable.HashSet[Any]): Any = {
  val array = child.dataType match {
    case BinaryType => buffer.iterator().map(_.asInstanceOf[UnsafeArrayData].toByteArray).toArray()
    case _ => buffer.toArray()
  }
  new GenericArrayData(array)
}